numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,444 @@
+     */
+    int len1 = context->descriptors[0]->elsize / sizeof(character);
+    int len2 = context->descriptors[1]->elsize / sizeof(character);
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *out = data[2];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        int cmp = string_cmp<rstrip>(
+                len1, (character *)in1, len2, (character *)in2);
+        npy_bool res;
+        if (comp == Py_EQ) {
+            res = cmp == 0;
+        }
+        else if (comp == Py_NE) {
+            res = cmp != 0;
+        }
+        else if (comp == Py_LT) {
+            res = cmp < 0;
+        }
+        else if (comp == Py_LE) {
+            res = cmp <= 0;
+        }
+        else if (comp == Py_GT) {
+            res = cmp > 0;
+        }
+        else if (comp == Py_GE) {
+            res = cmp >= 0;
+        }
+        else {
+            assert(0);
+        }
+        *(npy_bool *)out = res;
+
+        in1 += strides[0];
+        in2 += strides[1];
+        out += strides[2];
+    }
+    return 0;
+}
+
+
+/*
+ * Machinery to add the string loops to the existing ufuncs.
+ */
+
+/*
+ * This function replaces the strided loop with the passed in one,
+ * and registers it with the given ufunc.
+ */
+static int
+add_loop(
+        PyObject *umath, const char *ufunc_name,
+        PyArrayMethod_Spec *spec, PyArrayMethod_StridedLoop *loop)
+{
+    PyObject *name = PyUnicode_FromString(ufunc_name);
+    if (name == nullptr) {
+        return -1;
+    }
+    PyObject *ufunc = PyObject_GetItem(umath, name);
+    Py_DECREF(name);
+    if (ufunc == nullptr) {
+        return -1;
+    }
+    spec->slots[0].pfunc = (void *)loop;
+
+    int res = PyUFunc_AddLoopFromSpec(ufunc, spec);
+    Py_DECREF(ufunc);
+    return res;
+}
+
+
+extern "C" {
+    NPY_NO_EXPORT int
+    init_string_ufuncs(PyObject *umath);
+}
+
+NPY_NO_EXPORT int
+init_string_ufuncs(PyObject *umath)
+{
+    int res = -1;
+    /* NOTE: This should recieve global symbols? */
+    PyArray_DTypeMeta *String = PyArray_DTypeFromTypeNum(NPY_STRING);
+    PyArray_DTypeMeta *Unicode = PyArray_DTypeFromTypeNum(NPY_UNICODE);
+    PyArray_DTypeMeta *Bool = PyArray_DTypeFromTypeNum(NPY_BOOL);
+
+    /* We start with the string loops: */
+    PyArray_DTypeMeta *dtypes[] = {String, String, Bool};
+    /*
+     * We only have one loop right now, the strided one.  The default type
+     * resolver ensures native byte order/canonical representation.
+     */
+    PyType_Slot slots[] = {
+        {NPY_METH_strided_loop, nullptr},
+        {0, nullptr}
+    };
+
+    PyArrayMethod_Spec spec = {};
+    spec.name = "templated_string_comparison";
+    spec.nin = 2;
+    spec.nout = 1;
+    spec.dtypes = dtypes;
+    spec.slots = slots;
+    spec.flags = NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    /* Use this loop variable for typing more explicitly */
+    PyArrayMethod_StridedLoop *loop;
+
+    /* TODO: It would be nice to condense the below */
+    /* All String loops */
```

Comment:
:thinking: I have to admit, that is utterly non-obvious to me right now, I seem to be missing some concepts to make easy sense of the example (but I admit, I did not take more than 10 minutes to try to understand it).

Reply:
Ah, now I get it.  But to do this well, I would need to pass "tuples" with `"equal", Py_EQ`, I guess.  At least unless I double down on the "equal" spelling and use that everywhere rather than the int.

Reply:
The function `int_to_string` would create the binding between `comp` and the string name. If you prefer, I'll propose a patch once this one lands and I'll add you... as a reviewer ;-)

Reply:
Hmm, I half see it, but it would be awesome if we could do something like a clear mapping somewhere, maybe not spelled as a switch statement? If you are up to a followup, that is great :).